### PR TITLE
Add kubeval docs

### PR DESCRIPTION
**Pull request by mimotej**
This PR:

- [X] Updates contributing.md and adds information about `kustomize build` and `kubeval validation` steps
- [X] Adds docs for kubeval

### CODE REVIEW

File: docs/content/kubeval/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+This manifest will be properly build using `kustomize build` but `kubeval` will be able to catch this issue and report it back.
+
+## Usage
+
+We can use kubeval on manifests which are either stored in files or we can input manifests from stdin. This is useful in our case, because we can pipe output from `kustomize build` to kubeval and check if manifests are correct.
+
+By default `kubeval` uses default schemas which can be found [here](https://kubernetesjsonschema.dev). This is not ideal in our case because we have many CRDs which do not have schemas in default repository. For this reason we are generating our own schemas and using them in kubeval (more about how we are generating and updating schemas can be found [here](https://www.operate-first.cloud/apps/content/kubeval/updating_schema_store.html)). To use generated schemas we are using `--schema-location` option.
+
+We are also using `--strict` option which allows no additional properties in schema. Last option we specify is `--skip-kinds` which allows us to skip validation of CRDs which don't have valid schemas.
```
Comment (HumairAK) [inline docs/content/kubeval/README.md:26]: Let's link the actual script as well: https://github.com/operate-first/toolbox/blob/master/scripts/test-kubeval-validation

File: docs/content/kubeval/updating_schema_store.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Updating schemas in schema store
+
+Our `apps` repository uses CRDs for which schemas can't be found in default schema repository. This means that we have to use our own schema repository. For this we have used [script](https://github.com/sabre1041/k8s-manifest-validation/blob/main/scripts/build_schema.py) written by Andrew Block. This script generates schemas from OpenAPI specification using [openapi2jsonschema](https://github.com/instrumenta/openapi2jsonschema). We have also extended this script to generate `skip_kinds.txt` which contains names of the CRDs which don't have valid schemas and will not be validated in `kubeval` validation. Script can be found in [schema-store](https://github.com/operate-first/schema-store) repository.
+To use script you run have to login to cluster with `oc` and run following:
```
Comment (HumairAK) [inline docs/content/kubeval/updating_schema_store.md:4]:
```suggestion
To use this script you will have to login to the cluster with `oc` and run following:
```

File: contributing.md [hunk 3]
```diff
@@ -114,4 +114,14 @@ The following is a list of some common commands and their uses, but more can be
 
 ## Tests
 
+### Linting
+
 To run linting tests, we use [pre-commit](https://pre-commit.com/). Run `pre-commit install` after you clone the repo. Then, `pre-commit` will run automatically on git commit. If any one of the tests fail, add and commit the changes made by pre-commit. Once the pre-commit check passes, you can make your PR.
+
+### Kustomize build
+
+We are using `kustomize build` in our CI pipeline to test if syntax of overlays is correct. This command is run on changes which are found between commit in PR and latest commit in master branch in `apps` repository.
```
Comment (HumairAK) [inline contributing.md:123]:
```suggestion
We are using `kustomize build` in our CI pipeline to test if the syntax of overlays is correct. This command is run against changes which are found between commit in PR and latest commit in master branch in `apps` repository.
```

File: contributing.md [hunk 4]
```diff
@@ -114,4 +114,14 @@ The following is a list of some common commands and their uses, but more can be
...
+
+### Kubeval validation
+
+As the last step in our validation pipeline we are using `kubeval`. `Kubeval` checks if the manifests are valid against schemas which we are stored in [schema-store](https://github.com/operate-first/schema-store). These schemas were extracted from clusters using [script](https://github.com/operate-first/schema-store/blob/main/build_schema.py) which uses OpenAPI spec. More information about kubeval validation we are using can be found in [documentation](https://www.operate-first.cloud/apps/content/kubeval/README.html).
```
Comment (HumairAK) [inline contributing.md:127]:
```suggestion
As the last step in our validation pipeline we are using `kubeval`. `Kubeval` checks if the manifests are valid against schemas which we store in [schema-store](https://github.com/operate-first/schema-store). These schemas were extracted from clusters using [script](https://github.com/operate-first/schema-store/blob/main/build_schema.py) which uses OpenAPI spec. More information about how we are using kubeval validation can be found in [documentation](https://www.operate-first.cloud/apps/content/kubeval/README.html).
```

File: docs/content/kubeval/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Kubeval validation
+
+As a part of our CI pipelines in `apps` repository we are using kubeval to validate our schemas. Kubeval checks validity of the CRDs against the schemas stored in schema-store repository. If there is any error it is reported. This approach discovers more issues than just using `kustomize build` command. For example in following manifest `spec.port.targetPort` is not valid:
```
Comment (HumairAK) [inline docs/content/kubeval/README.md:3]:
```suggestion
As a part of our CI pipelines in `apps` repository we are using kubeval to validate our schemas. Kubeval checks validity of the CRDs against the schemas stored in schema-store repository. If there are any errors, it is reported. This approach will check for api errors that can sneak through the `kustomize build` command. For example in following manifest `spec.port.targetPort` is not valid:
```

File: docs/content/kubeval/updating_schema_store.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+```bash
+python build_schema.py -u $(oc whoami --show-server) -t $(oc whoami -t) -s STRICT
+```
+
+Script generates folder called `openshift-json-schema/master-standalone-strict` which contains schemas for all CRDs which are present in the cluster.
```
Comment (HumairAK) [inline docs/content/kubeval/updating_schema_store.md:10]:
```suggestion
This script generates a folder called `openshift-json-schema/master-standalone-strict` which contains schemas for all CRDs which are present in this cluster.
```

File: docs/content/kubeval/updating_schema_store.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+Script is then used in github actions, where we [run cronjob](https://github.com/operate-first/schema-store/blob/main/.github/workflows/update-schemas.yaml), which generates schemas from each cluster and joins them together. In the end if the PR doesn't exist it is created with new changes or if there is opened PR changes are updated.
```
Comment (HumairAK) [inline docs/content/kubeval/updating_schema_store.md:11]:
```suggestion
This script is then used in github actions, where we [run cronjob](https://github.com/operate-first/schema-store/blob/main/.github/workflows/update-schemas.yaml), which generates schemas from each cluster and joins them together. In the end if the PR doesn't exist it is created with new changes or if there is opened PR changes are updated.
```

File: docs/content/kubeval/README.md (hunk 1 above)
Comment (mimotej) [reply docs/content/kubeval/README.md:26]: Resolved in latest commit.

File: docs/content/kubeval/updating_schema_store.md (hunk 2 above)
Comment (mimotej) [reply docs/content/kubeval/updating_schema_store.md]: Resolved in latest commit.

File: contributing.md (hunk 4 above)
Comment (mimotej) [reply contributing.md]: Resolved in latest commit.

File: docs/content/kubeval/README.md (hunk 5 above)
Comment (mimotej) [reply docs/content/kubeval/README.md]: Resolved in latest commit.

File: docs/content/kubeval/updating_schema_store.md (hunk 6 above)
Comment (mimotej) [reply docs/content/kubeval/updating_schema_store.md]: Resolved in latest commit.

File: docs/content/kubeval/updating_schema_store.md (hunk 7 above)
Comment (mimotej) [reply docs/content/kubeval/updating_schema_store.md]: Resolved in latest commit.

File: contributing.md (hunk 3 above)
Comment (mimotej) [reply contributing.md]: Resolved in latest commit.

**Comment by mimotej** [comment]
@HumairAK Fixed all issues thx :+1:

**Comment by Gregory-Pereira** [comment]
For some reason, the kustomize-build check isn't running against these doc changes. The kubeval cehck istn running either, coudl be because a rebase is in order. Il try to checkout the Kustomize-build check.
/retest

**Comment by Gregory-Pereira** [comment]
Seems the retest, is not being picked up by logs, trying test-all.
/test-all

**Comment by Gregory-Pereira** [comment]
Its getting picked up by prow hook but the logs seem to fail to trigger the tests, or maybe have correctly identified not to trigger them,  ```
{"author":"Gregory-Pereira","client":"github","component":"hook","duration":"252.524656ms","event-GUID":"83187570-2718-11ed-8bd7-223c116faa98","event-type":"issue_comment","file":"prow/github/client.go:891","func":"k8s.io/test-infra/prow/github.(*client).log.func2","level":"debug","msg":"GetPullRequestChanges(operate-first, apps, 2273) finished","org":"operate-first","plugin":"trigger","pr":2273,"repo":"apps","severity":"debug","time":"2022-08-28T21:30:18Z","url":"https://github.com/operate-first/apps/pull/2273#issuecomment-1229557978"}
{"author":"Gregory-Pereira","component":"hook","event-GUID":"83187570-2718-11ed-8bd7-223c116faa98","event-type":"issue_comment","file":"prow/pjutil/filter.go:135","func":"k8s.io/test-infra/prow/pjutil.FilterPresubmits","level":"debug","msg":"Filtered 3 jobs, found 0 to trigger","org":"operate-first","plugin":"trigger","pr":2273,"repo":"apps","severity":"debug","time":"2022-08-28T21:30:18Z","to-trigger":null,"url":"https://github.com/operate-first/apps/pull/2273#issuecomment-1229557978"}
{"args":["gc.auto","0"],"client":"git","component":"hook","file":"prow/git/git.go:485","func":"k8s.io/test-infra/prow/git.(*Repo).Config","level":"info","msg":"Running git config.","severity":"info","time":"2022-08-28T21:31:27Z"} ```

**Comment by Gregory-Pereira** [comment]
Interesting, when querying test directly, it will run it despite not being able to trigger it before. I am assuming this is not a flake, as the same thing is the case with the operator-docs pr. Rebase should fix the `kustomize-build`, in the mean time manually triggering kubeval test.
/test kubeval-validation

**Comment by tumido** [comment]
/override kubeval-validation
/override kustomize-build

Honestly this PR shouldn't be even checked for kubeval and kustomize. Don't bother with it, let's rather fix Prow to properly skip this PR.

**Pull request merged**